embulk-core/src/main/java/org/embulk/spi/TempFileSpace.java

```diff
@@ -7,58 +7,135 @@
+     *
+     * <p>It creates {@link TempFileSpace} under the specified {@code baseDir} with the {@code prefix}.
+     */
+    public static TempFileSpace with(final Path baseDir, final String prefix) throws IOException {
+        if (baseDir == null || prefix == null) {
+            throw new IllegalArgumentException("TempFileSpace is created with null.");
+        }
+        if (!baseDir.isAbsolute()) {
+            throw new IllegalArgumentException("TempFileSpace is created under a relative path.");
         }
-        this.dir = dir;
+        if (!Files.isDirectory(baseDir)) {  // Following symlinks -- no LinkOption.NOFOLLOW_LINKS.
+            throw new IOException("TempFileSpace is created under non-directory.");
+        }
+
+        return new TempFileSpace(true, baseDir, prefix, null);
     }
 
+    /**
+     * Creates a temporary file with the default file name prefix, and the default file extension suffix.
+     *
+     * @return A temporary {@link java.io.File} created
+     */
     public File createTempFile() {
         return this.createTempFile("tmp");
     }
 
+    /**
+     * Creates a temporary file with the default file name prefix, and the specified file extension suffix.
+     *
+     * @param fileExt  The file extension suffix without a leading dot ({@code '.'})
+     * @return A temporary {@link java.io.File} created
+     */
     public File createTempFile(final String fileExt) {
         // Thread names contain ':' which is not valid as file names in Windows.
         return this.createTempFile(Thread.currentThread().getName().replaceAll(":", "_") + "_", fileExt);
     }
 
+    /**
+     * Creates a temporary file with the specified file name prefix, and the specified file extension suffix.
+     *
+     * @param prefix  The file name prefix
+     * @param fileExt  The file extension suffix without a leading dot ({@code '.'})
+     * @return A temporary {@link java.io.File} created
+     */
     public synchronized File createTempFile(final String prefix, final String fileExt) {
         try {
-            if (!this.dirCreated) {
-                this.dir.mkdirs();
-                this.dirCreated = true;
-            }
-            return File.createTempFile(prefix, "." + fileExt, this.dir);
+            this.createTempDirectoryIfRequired();
+
+            final Path tempFile = Files.createTempFile(this.tempDirectoryCreated.get(), prefix, "." + fileExt);
+            logger.debug("TempFile \"{}\" is created.", tempFile);
+            return tempFile.toFile();
         } catch (final IOException ex) {
             throw new TempFileException(ex);
         }
     }
 
+    /**
+     * Cleans up the temporary file space, and everything in the space.
+     *
+     * <p>It is to be called along with Embulk's cleanup process. Plugins should not call this directly.
+     */
     public synchronized void cleanup() {
```

Comment:
`Plugins should not call this directly` <- can we make it package-private?

Reply:
Possibly, but this method has been `public` for long, and we may call this out of this package even while in `embulk-core`.